### PR TITLE
- remove separate BQ cutoff for het sensitivity in WGS metrics. This …

### PR DESCRIPTION
**Pull request by yfarjoun**
…causes incorrect depth and therefore sensitivity results whenever it's relevant. Also, the 17 cutoff in GATK onl refers to the graph building, for the pair HMM there is no minimum quality.

- added a test to verify that the base-based exclusions affect the coverage and base-quality distributions correctly

### CODE REVIEW

**Comment by yfarjoun** [comment]
@nh13 do you mind reviewing this?

**Comment by yfarjoun** [comment]
fixed test.

File: src/java/picard/analysis/CollectWgsMetrics.java [hunk 1]
```diff
@@ -318,14 +317,10 @@ public void addInfo(final SamLocusIterator.LocusInfo info, final ReferenceSequen
...
-            int pileupSizeForBaseQHetSens = 0;
             for (final SamLocusIterator.RecordAndOffset recs : info.getRecordAndPositions()) {
-                pileupSizeForBaseQHetSens++;
-                if(pileupSizeForBaseQHetSens <= coverageCap) {
-                    baseQHetSensHistogram[recs.getRecord().getBaseQualities()[recs.getOffset()]]++;
-                }
 
-                if (recs.getBaseQuality() < MINIMUM_BASE_QUALITY)                   { ++basesExcludedByBaseq;   continue; }
+                if (recs.getBaseQuality() < MINIMUM_BASE_QUALITY ||
+                        SequenceUtil.isNoCall(recs.getReadBase()))                  { ++basesExcludedByBaseq;   continue; }
```
Comment (nh13) [inline src/java/picard/analysis/CollectWgsMetrics.java]: `PCT_EXC_BASEQ` should be updated to say that no call bases are also excluded, sine what if I have a Q40 `N`?

File: src/java/picard/analysis/CollectWgsMetrics.java [hunk 2]
```diff
@@ -334,9 +329,9 @@ public void addInfo(final SamLocusIterator.LocusInfo info, final ReferenceSequen
                 }
             }
 
-            final int depth = Math.min(readNames.size(), coverageCap);
-            if (depth < readNames.size()) basesExcludedByCapping += readNames.size() - coverageCap;
-            histogramArray[depth]++;
+            final int depth = Math.min(pileupSize, coverageCap);
+            if (depth < pileupSize) basesExcludedByCapping += pileupSize - coverageCap;
+            depthHistogramArray[depth]++;
```
Comment (nh13) [inline src/java/picard/analysis/CollectWgsMetrics.java]: I think `readNames.size()` should equal `pileupSize`.  Is that correct?

File: src/java/picard/analysis/CollectWgsMetrics.java [hunk 3]
```diff
@@ -377,12 +363,21 @@ protected void addMetricsToFile(final MetricsFile<WgsMetrics, Integer> file,
             file.addHistogram(depthHistogram);
         }
 
-        protected Histogram<Integer> getDepthHistogram() {
-            final Histogram<Integer> depthHistogram = new Histogram<>("coverage", "count");
-            for (int i=0; i<histogramArray.length; ++i) {
-                depthHistogram.increment(i, histogramArray[i]);
+        protected Histogram<Integer> getDepthHistogram(){
```
Comment (nh13) [inline src/java/picard/analysis/CollectWgsMetrics.java]: `){` needs more spaces, here and below.

File: src/java/picard/analysis/CollectWgsMetrics.java [hunk 4]
```diff
@@ -377,12 +363,21 @@ protected void addMetricsToFile(final MetricsFile<WgsMetrics, Integer> file,
...
+            return getHistogram(depthHistogramArray,"coverage", "count");
+        }
+
+        protected Histogram<Integer> getBaseQHistogram(){
+            return getHistogram(baseQHistogramArray, "value", "baseq_count");
+        }
+
+
+        private Histogram<Integer> getHistogram(final long[] array, final String binLabel, final String valueLabel) {
```
Comment (nh13) [inline src/java/picard/analysis/CollectWgsMetrics.java]: some new lines above

File: src/tests/java/picard/analysis/CollectWgsMetricsFromSampledSitesTest.java [hunk 5]
```diff
@@ -75,7 +76,7 @@ public void testOnePos() throws IOException {
             Assert.assertEquals(metrics.PCT_EXC_DUPE, 0.181818); // 2 of 11
             Assert.assertEquals(metrics.PCT_EXC_UNPAIRED, 0.090909); // 1 of 9
             Assert.assertEquals(metrics.PCT_EXC_BASEQ, 0.090909); // 1 of 9
-            Assert.assertEquals(metrics.HET_SNP_SENSITIVITY, 0.34655, .02);
+            Assert.assertEquals(metrics.HET_SNP_SENSITIVITY, 0.4955, .02);
```
Comment (nh13) [inline src/tests/java/picard/analysis/CollectWgsMetricsFromSampledSitesTest.java]: 🅰️ +

**Comment by nh13** [comment]
@yfarjoun back to you

File: src/java/picard/analysis/CollectWgsMetrics.java (hunk 2 above)
Comment (yfarjoun) [inline src/java/picard/analysis/CollectWgsMetrics.java]: yes, there is no bug here (despite rumors to the contrary...) just an attempt at clarifying.

**Comment by yfarjoun** [comment]
thanks for the review @nh13 responding here.

File: src/java/picard/analysis/CollectWgsMetrics.java [hunk 6]
```diff
@@ -99,7 +99,8 @@
     @Option(shortName = "MQ", doc = "Minimum mapping quality for a read to contribute coverage.", overridable = true)
     public int MINIMUM_MAPPING_QUALITY = 20;
 
-    @Option(shortName = "Q", doc = "Minimum base quality for a base to contribute coverage.", overridable = true)
+    @Option(shortName = "Q", doc = "Minimum base quality for a base to contribute coverage. N's in the SEQ will be excluded" +
+            "from coverage and counted as though they are excluded for quality regardless of the value of this parameter.", overridable = true)
```
Comment (nh13) [inline src/java/picard/analysis/CollectWgsMetrics.java]: This was hard to read.

'N's will be counted as having a base quality of negative infinity and will be excluded from coverage.`?

**Comment by nh13** [comment]
@yfarjoun minor comment on wording, but since English is not the first language of either of us, probably ask someone else.

**Comment by nh13** [comment]
@yfarjoun 👍  I think if we both got together we may be able to write good.

**Pull request merged**